planner/core/optimizer.go

```diff
@@ -373,10 +374,143 @@ func postOptimize(sctx sessionctx.Context, plan PhysicalPlan) PhysicalPlan {
 	mergeContinuousSelections(plan)
 	plan = eliminateUnionScanAndLock(sctx, plan)
 	plan = enableParallelApply(sctx, plan)
+	handleFineGrainedShuffle(sctx, plan)
 	checkPlanCacheable(sctx, plan)
 	return plan
 }
 
+// Only for MPP(Window<-[Sort]<-ExchangeReceiver<-ExchangeSender).
+// TiFlashFineGrainedShuffleStreamCount:
+// == 0: fine grained shuffle is disabled.
+// > 0: use TiFlashFineGrainedShuffleStreamCount as stream count.
+// < 0: use TiFlashMaxThreads as stream count when it's greater than 0. Otherwise use DefStreamCountWhenMaxThreadsNotSet.
+func handleFineGrainedShuffle(sctx sessionctx.Context, plan PhysicalPlan) {
+	streamCount := sctx.GetSessionVars().TiFlashFineGrainedShuffleStreamCount
+	if streamCount == 0 {
+		return
+	}
+	if streamCount < 0 {
+		if sctx.GetSessionVars().TiFlashMaxThreads > 0 {
+			streamCount = sctx.GetSessionVars().TiFlashMaxThreads
+		} else {
+			streamCount = variable.DefStreamCountWhenMaxThreadsNotSet
+		}
+	}
+	setupFineGrainedShuffle(uint64(streamCount), plan)
+}
+
+func setupFineGrainedShuffle(streamCount uint64, plan PhysicalPlan) {
+	if tableReader, ok := plan.(*PhysicalTableReader); ok {
+		if _, isExchangeSender := tableReader.tablePlan.(*PhysicalExchangeSender); isExchangeSender {
+			helper := fineGrainedShuffleHelper{shuffleTarget: unknown, plans: make([]*basePhysicalPlan, 1)}
+			setupFineGrainedShuffleInternal(tableReader.tablePlan, &helper, streamCount)
+		}
+	} else {
+		for _, child := range plan.Children() {
+			setupFineGrainedShuffle(streamCount, child)
+		}
+	}
+}
+
+type shuffleTarget uint8
+
+const (
+	unknown shuffleTarget = iota
+	window
+	agg
+	joinBuild
+	joinProbe
+)
+
+type fineGrainedShuffleHelper struct {
+	shuffleTarget shuffleTarget
+	plans         []*basePhysicalPlan
+}
+
+func (h *fineGrainedShuffleHelper) clear() {
+	h.shuffleTarget = unknown
+	h.plans = h.plans[:0]
+}
+
+func (h *fineGrainedShuffleHelper) add(t shuffleTarget, p *basePhysicalPlan) {
+	h.shuffleTarget = t
+	h.plans = append(h.plans, p)
+}
+
+func setupFineGrainedShuffleInternal(plan PhysicalPlan, helper *fineGrainedShuffleHelper, streamCount uint64) {
+	switch x := plan.(type) {
+	case *PhysicalWindow:
+		if len(x.PartitionBy) > 0 {
```

Comment:
I think there is no need to check the partition by column here, because if there is not partition by, then the exchange type must be `Passthrough`, which will be handled in `ExchangeSender`

Reply:
Looks like the following plan may not work. Because `Window_10` has no partition by, but `Exchange_16` is still hash partition.
```
MySQL [test]> explain select row_number() over w2, row_number() over w1 from t1 window w2 as (order by c1), w1 as (partition by c1);
+--------------------------------------+----------+--------------+---------------+-----------------------------------------------------------------------------------------------+
| id                                   | estRows  | task         | access object | operator info                                                                                 |
+--------------------------------------+----------+--------------+---------------+-----------------------------------------------------------------------------------------------+
| Projection_9                         | 10000.00 | root         |               | Column#6, Column#5                                                                            |
| └─TableReader_22                     | 10000.00 | root         |               | data:ExchangeSender_21                                                                        |
|   └─ExchangeSender_21                | 10000.00 | mpp[tiflash] |               | ExchangeType: PassThrough                                                                     |
|     └─Window_10                      | 10000.00 | mpp[tiflash] |               | row_number()->Column#6 over(order by test.t1.c1 rows between current row and current row)     |
|       └─Window_12                    | 10000.00 | mpp[tiflash] |               | row_number()->Column#5 over(partition by test.t1.c1 rows between current row and current row) |
|         └─Sort_17                    | 10000.00 | mpp[tiflash] |               | test.t1.c1                                                                                    |
|           └─ExchangeReceiver_16      | 10000.00 | mpp[tiflash] |               |                                                                                               |
|             └─ExchangeSender_15      | 10000.00 | mpp[tiflash] |               | ExchangeType: HashPartition, Hash Cols: [name: test.t1.c1, collate: binary]                   |
|               └─TableFullScan_14     | 10000.00 | mpp[tiflash] | table:t1      | keep order:false, stats:pseudo                                                                |
+--------------------------------------+----------+--------------+---------------+-----------------------------------------------------------------------------------------------+
9 rows in set (0.001 sec)
```

Reply:
I believe it is a wrong plan, after window_12, there should be a `Passthrough` exchange sender/receiver to collect all the data into one TiFlash node, @LittleFall PTAL

Reply:
Yeah, it is.

Reply:
Done. Removed and comments added.

Reply:
thanks for pointing, I will fix it